Warning log callback in client-go

mlavacca opened the pull request:
This PR introduces a callback for custom logging to use in the case of `ClientConfigLoadingRules.WarnIfAllMissing` is `true` and there are missing files.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

**What this PR does / why we need it**:

The `ClientConfigLoadingRules` fields have been enriched with a new callback to allow the customization of the logging function to use in the case of `WarnIfAllMissing` field is true and there are missing files. This approach allows both to specify a logging function different from `klog.Warningf` and not to log. 

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #94428

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The `klog.Warningf` in `ClientConfigLoadingRules.Load` has been replaced by a call of the exported callback field `ClientConfigLoadingRules.WarnIfAllMissingCallback` that is configured by default to `klog.Warningf`, but can be replaced with any `func(string, ...interface{})` or nil if no log is desired. (#94428, @smarterclayton)
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

mlavacca commented:
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
>  **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](mailto:login-issues@jira.linuxfoundation.org)
> 
> Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md). If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).

mlavacca commented:
/check-cla

mlavacca commented:
/check-cla

mlavacca commented:
I signed it

mlavacca commented:
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
>  **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](mailto:login-issues@jira.linuxfoundation.org)

I signed it

mlavacca commented:
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
>  **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](mailto:login-issues@jira.linuxfoundation.org)

/check-cla

mlavacca commented:
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
>  **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](mailto:login-issues@jira.linuxfoundation.org)

/check-cla

MHBauer commented:
/sig cli
/ok-to-test

mlavacca commented:
Hi @soltysh, @yliaog, could you please review this PR?

yliaog commented:
looks good, could you add a test case for this change?

fedebongio commented:
/assign @yliaog 
/triage accepted

mlavacca commented:
> looks good, could you add a test case for this change?

Yep, test added

yliaog commented on staging/src/k8s.io/client-go/tools/clientcmd/loader_test.go:
```diff
@@ -131,6 +132,20 @@ func TestToleratingMissingFiles(t *testing.T) {
+	_, err = loadingRules.Load()
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
```
there should be assertion about the warning logs produced as expected

mlavacca replied:
I implemented the assertion of the logging output through the klog library. I encountered some problems with bazel while importing the klog library in the testing package (this is my very first contribution to k8s, I'm learning how the contribution ecosystem is built) and I'm working to solve it.

mlavacca replied:
ok, solved. It should be ok.

yliaog commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -154,9 +158,10 @@ func NewDefaultClientConfigLoadingRules() *ClientConfigLoadingRules {
+		Precedence:               chain,
+		MigrationRules:           currentMigrationRules(),
+		WarnIfAllMissing:         warnIfAllMissing,
+		WarnIfAllMissingCallback: klog.V(4).Infof,
```
previously it was klog.Warningf, why switching to klog.V(4).Infof?

mlavacca replied:
I did not want to commit (I was just checking that `klog.V(x).Infof` works as well and I committed that line with no purpose). Restored to `klog.Warningf`.

mlavacca commented:
/reopen

mlavacca committed:
Warning log callback in client-go loading rules

This PR introduces a callback for custom logging to use in the case of
ClientConfigLoadingRules.WarnIfAllMissing is true and there are missing
files. A specific test checking the expected log has been added.

mlavacca commented:
/reopen

yliaog commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -219,8 +223,8 @@ func (rules *ClientConfigLoadingRules) Load() (*clientcmdapi.Config, error) {
 
-	if rules.WarnIfAllMissing && len(missingList) > 0 && len(kubeconfigs) == 0 {
-		klog.Warningf("Config not found: %s", strings.Join(missingList, ", "))
+	if rules.WarnIfAllMissing && len(missingList) > 0 && len(kubeconfigs) == 0 && rules.WarnIfAllMissingCallback != nil {
```
when the caller does not use NewDefaultClientConfigLoadingRules(), and does not set rules.WarnIfAllMissingCallback explicitly, it would be nil. In that case, it should default to klog.Warningf, which is the previous behavior.

mlavacca replied:
I implemented this behavior by means of the unexported variable `ClientConfigLoadingRules.warningMuted`. In addition, the field `ClientConfigLoadingRules.warnIfAllMissingCallback` has been unexported and made customizable by using a setter method.

yliaog commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -131,6 +131,12 @@ type ClientConfigLoadingRules struct {
 	WarnIfAllMissing bool
+
+	// WarnIfAllMissingCallback is the warning log callback to use in case of missing files.
+	warnIfAllMissingCallback func(string, ...interface{})
```
since all other fields in this struct are exported, why not exporting this too?

mlavacca replied:
The reason is strictly related to the other comment: by using the new variable `warningMuted` for distinguish between * `WarnIfAllMissingCallback` is nil and we log by default, and
* `WarnIfAllMissingCallback` is nil and we suppress the log

By letting it exported, it is possible to create a new inconsisent state: `WarnIfAllMissingCallback != nil` and `warningMuted == true`.

For this reason I provided the setter method through which is only possible to have 3 states:
* `WarnIfAllMissingCallback == nil` and `warningMuted == false` (explicit allocation)
*  `WarnIfAllMissingCallback == nil` and `warningMuted == true` (any allocation followed by `MuteWarningMissingList `)
* `WarnIfAllMissingCallback != nil` and `warningMuted == false` (default allocation or explicit allocation followed by a `SetWarnIfAllMissingCallback`)

yliaog commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -131,6 +131,12 @@ type ClientConfigLoadingRules struct {
+	warnIfAllMissingCallback func(string, ...interface{})
+
+	// warningMuted mutes the warning callback triggered in case of missing files
+	warningMuted bool
```
is it necessary? in the case someone does not want to log anything, he can provide a WarnIfAllMissingCallback that does nothing.

mlavacca replied:
Since the behavior that we want to implement is: if we allocate a new `ClientConfigLoadingRules`, no matter how (if through ClientConfigLoadingRules or by allocating it explicitly), it must be defaulted to `klog.Warningf`, then how can we distinguish the following cases between each oher?
1. the `ClientConfigLoadingRules` has been created by `NewDefaultClientConfigLoadingRules`, then we muted the log by assigning nil to `WarnIfAllMissingCallback` (we do not want lo log anthing).
2. The `ClientConfigLoadingRules` has been allocated explicitly without setting anything for `WarnIfAllMissingCallback`
 (we want to log with the default function).
In the above cases `WarnIfAllMissingCallback==nil` has different meaning, then I needed a parameter to distinguish between each other.

mlavacca replied:
Ok sorry, I did not read carefully. If you think that it would be better to use that approach, I can implement this simpler mechanism:
* If `WarnIfAllMissingCallback` is nil, then  log to `klog.Warningf`
* else log to `WarnIfAllMissingCallback`.

And it is up to who uses `ClientConfigLoadingRules` to provide a method that does nothing. 
I think that the approach I proposed is clearer, but I understand that it can be too much code for a very limited functional change.

mlavacca commented:
@yliaog I pushed a commit with the modification you proposed in the last review (`WarnIfAllMissingCallback` is manually set as a function that does nothing in case no log is desired). I can either squash or delete it depending on which solution you think is better.

yliaog commented:
thanks for the pr

/lgtm

mlavacca commented:
/assign @smarterclayton

mlavacca commented:
pinging @smarterclayton for approval

mlavacca commented:
hi @smarterclayton, can you please take a look at this PR?

mlavacca commented:
/unassign @smarterclayton

mlavacca commented:
/assign @liggitt

liggitt commented:
/unassign
/assign @smarterclayton 

will defer to clayton as the author of the issue this is fixing

soltysh commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -131,6 +131,9 @@ type ClientConfigLoadingRules struct {
 	WarnIfAllMissing bool
+
+	// WarnIfAllMissingCallback is the warning log callback to use in case of missing files.
+	WarnIfAllMissingCallback func(string, ...interface{})
```
I'd probably go with something like:
```golang
WarningHandler func(string)
```

mlavacca replied:
Without the formatting directives?

soltysh replied:
yeah, anyone willing to format they'll use `fmt.Sprintf`

soltysh commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -220,7 +223,13 @@ func (rules *ClientConfigLoadingRules) Load() (*clientcmdapi.Config, error) {
-		klog.Warningf("Config not found: %s", strings.Join(missingList, ", "))
+		missingString := strings.Join(missingList, ", ")
+
+		if rules.WarnIfAllMissingCallback != nil {
```
Here if no `WarbubgHandler` is defined we'd silently ignore the problem. The original issue was to entirely get rid of warnings from library code.

liggitt replied:
is kubectl able to set the warning handler to preserve visibility to the failure when running with kubectl? Shouldn't we be setting that?

soltysh replied:
yeah, it should be possible to do, that's why I'm advocating having this be settable.

mlavacca replied:
Should I use `klog.Infof` by default instead of `klog.Warningf`?

mlavacca commented:
/retest

soltysh commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -220,7 +223,13 @@ func (rules *ClientConfigLoadingRules) Load() (*clientcmdapi.Config, error) {
+		if rules.WarningHandler != nil {
+			rules.WarningHandler(fmt.Sprintf("Config not found: %s", missingString))
+		} else {
+			klog.Warningf("Config not found: %s", missingString)
```
Don't log, if `WarningHandler` is nil, just silently ignore the problem.

soltysh replied:
Alternatively, you can use `klog.V(1).Infof`.

soltysh commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -131,6 +131,9 @@ type ClientConfigLoadingRules struct {
 	WarnIfAllMissing bool
+
+	// WarningHandler is the warning log callback to use in case of missing files.
+	WarningHandler func(string)
```
I'd like this being set in https://github.com/kubernetes/kubernetes/blob/f6152d152140f42815765c7d9703b9817943c345/staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go#L132 which will ensure this is being used and actually consumed by kubectl itself.

soltysh commented:
/assign

soltysh reviewed (approved):
this lgtm, but you need to rebase

mlavacca commented:
/test pull-kubernetes-node-e2e

mlavacca commented:
> this lgtm, but you need to rebase

Done. I don't know why verify test still doesn't pass

ohbus reviewed (approved):
/lgtm

yliaog commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -129,6 +129,20 @@ type ClientConfigLoadingRules struct {
 	WarnIfAllMissing bool
+
+	// Warner is the warning log callback to use in case of missing files.
+	Warner WarningHandler
```
i think WarningHandler is easier to read than Warner

mlavacca replied:
Do you suggest to use WarningHandler as instance and Warner as type?

yliaog replied:
what about using: WarningHandler WarningHandlerFunc

mlavacca committed:
warningHandler renamed to WarninghandlerFunc

mlavacca commented:
/retest

mlavacca commented:
/test pull-kubernetes-e2e-kind

yliaog commented on staging/src/k8s.io/client-go/tools/clientcmd/loader_test.go:
```diff
@@ -121,14 +123,57 @@ func TestNonExistentCommandLineFile(t *testing.T) {
+
+	expectedLog := fmt.Sprintf("Config not found: %s", envVarvalue)
+	if !strings.Contains(buffer.String(), expectedLog) {
+		t.Fatalf("expected log: \"%s\"", expectedLog)
```
i think the 'actualLog' (buffer.String()) is more interesting than expectedLog

yliaog commented on staging/src/k8s.io/client-go/tools/clientcmd/loader_test.go:
```diff
@@ -121,14 +123,57 @@ func TestNonExistentCommandLineFile(t *testing.T) {
+
+	expectedLog := fmt.Sprintf("Config not found: %s", envVarvalue)
+	if !strings.Contains(buffer.String(), expectedLog) {
+		t.Fatalf("expected log: \"%s\"", expectedLog)
```
ditto

mlavacca committed:
test fail messages changed

mlavacca commented:
/test pull-kubernetes-e2e-kind

yliaog commented:
/lgtm
